prow/gerrit/gerrit.go

```diff
@@ -0,0 +1,189 @@
+func NewController(instance string, projects []string, kc *kube.Client, logger *logrus.Entry, ca *config.Agent) (*Controller, error) {
+	c, err := gerrit.NewClient(instance, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	if logger == nil {
+		logger = logrus.NewEntry(logrus.StandardLogger())
+	}
+
+	return &Controller{
+		instance:   instance,
+		projects:   projects,
+		kc:         kc,
+		log:        logger,
+		ca:         ca,
+		gc:         c.Changes,
+		lastUpdate: time.Now(),
+	}, nil
+}
+
+// sync looks for newly made gerrit changes
+// and creates prowjobs according to presubmit specs
+func (c *Controller) Sync(rateLimit int) error {
+	syncTime := time.Now()
+	changes, err := c.QueryChanges(rateLimit)
+	if err != nil {
+		return fmt.Errorf("failed query changes : %v", err)
+	}
+
+	for _, change := range changes {
+		if err := c.ProcessChange(change); err != nil {
+			c.log.WithError(err).Error("Failed process change %v", change.CurrentRevision)
+		}
+	}
+
+	c.lastUpdate = syncTime
+	c.log.Infof("Processed %d changes", len(changes))
+	return nil
+}
+
+// QueryChanges will query all gerrit changes since controller's last sync loop
+func (c *Controller) QueryChanges(rateLimit int) (map[string]gerrit.ChangeInfo, error) {
+	opt := &gerrit.QueryChangeOptions{}
+	opt.Query = []string{}
+	for _, proj := range c.projects {
+		opt.Query = append(opt.Query, "project:"+proj)
+	}
+	opt.AdditionalFields = []string{"CURRENT_REVISION"}
+
+	done := false
+
+	// store a map of changeID:change
+	pending := map[string]gerrit.ChangeInfo{}
+	start := 0
+
+	for !done {
+		opt.Limit = rateLimit
+		opt.Start = start
+
+		// The change output is sorted by the last update time, most recently updated to oldest updated.
+		// Gerrit API docs: https://gerrit-review.googlesource.com/Documentation/rest-api-changes.html#list-changes
+		changes, _, err := c.gc.QueryChanges(opt)
+		if err != nil {
+			// should not happen? Let next sync loop catch up
+			return pending, fmt.Errorf("failed to query gerrit changes: %v", err)
+		}
+
+		if len(*changes) == 0 {
+			break
+		}
+
+		for _, change := range *changes {
+			// if we already processed this change, then we stop the current sync loop
+			layout := "2006-01-02 15:04:05"
+			updated, err := time.Parse(layout, change.Updated)
+			if err != nil {
+				c.log.WithError(err).Error("Parse time %v failed", change.Updated)
+				start++
+				continue
+			}
+
+			// process if updated later than last updated
+			// stop if already parsed
+			if updated.After(c.lastUpdate) {
+				// in case of a new change comes in during the current sync cycle
+				if _, ok := pending[change.CurrentRevision]; !ok {
+					pending[change.CurrentRevision] = change
+				}
+				start++
+			} else {
+				done = true
+				break
+			}
+		}
+	}
+
+	return pending, nil
+}
+
+// ProcessChange creates new presubmit prowjobs base off the gerrit changes
+func (c *Controller) ProcessChange(change gerrit.ChangeInfo) error {
```

Comment:
In order to handle restarts properly, this function will need to check for existing jobs and only start a new job if it hasn't been started already. (This is a TODO in the PR body.)

However, because we are querying for all changes on Gerrit after a restart, if we just list ProwJobs to look for previous runs we will end up retriggering jobs for older changes because the ProwJob records for the previous runs will have been GC'ed by sinker. I'm not sure what the best way to handle this is. We could make the Gerrit controller ignore changes older than sinker's max prowjob age, but that seems kind of gross and limiting.

Reply:
Currently upon restart, lastUpdated will start with `time.Now()`, and should not result in duplicated triggers.

How does trigger handles this? I didn't see a similar logic in pr.go, I thought Plank would handle no trigger the already started presubmit job?

Reply:
`terminateDupes` deletes old jobs when a new one is intentionally retriggered. It isn't related to preventing a job from being retriggered improperly.
Plank doesn't suffer from this problem because it is event based instead of polling based so we won't retrigger tests for old PRs even if the prowjobs for the PR have been garbage collected because we won't rereceive events for the PR.

Reply:
I think I can save the last updated time per sync loop, and restore from there on restart, and it should poll new changes properly